README.md
Money
=====

Generate Docs with [ApiGen](http://www.apigen.org/)
--------
```php
composer man
```

Testing
--------
```php
composer test
```

---

PHP 5.3+ library to make working with money safer, easier, and fun!

> "If I had a dime for every time I've seen someone use FLOAT to store currency, I'd have $999.997634" -- [Bill Karwin](https://twitter.com/billkarwin/status/347561901460447232)

In short: You shouldn't represent monetary values by a float. Wherever
you need to represent money, use this Money value object.

```php
<?php

use Money\Money;

$fiveEur = Money::EUR(500);
$tenEur = $fiveEur->add($fiveEur);

list($part1, $part2, $part3) = $tenEur->allocate(array(1, 1, 1));
assert($part1->equals(Money::EUR(334)));
assert($part2->equals(Money::EUR(333)));
assert($part3->equals(Money::EUR(333)));
```

The documentation is available at http://money.readthedocs.org


Installation
------------

Install the library using [composer][1]. Add the following to your `composer.json`:

```json
{
    "require": {
        "mathiasverraes/money": "dev-master"
    },
    "minimum-stability": "dev"
}
```

Now run the `install` command.

```sh
$ composer.phar install
```

Integration
-----------

See [`MoneyBundle`][2] or [`TbbcMoneyBundle`][4] for [Symfony integration][3] or [`DoctrineMoneyModule`][5] for [`Zend Framework 2`][6].

[1]: http://getcomposer.org/
[2]: https://github.com/pink-tie/MoneyBundle/
[3]: http://symfony.com/
[4]: https://github.com/TheBigBrainsCompany/TbbcMoneyBundle
[5]: https://github.com/zfbrasil/doctrine-money-module
[6]: http://framework.zend.com/


CHANGELOG.md
Money Changelog
===============

- 2014-03-22 Removed \Money\InvalidArgumentException in favour of plain old InvalidArgumentException
- 2014-03-22 Introduce RoundingMode object, used to specify desired rounding
- 2014-03-22 Introduced RoundingMode backwards compatible API changes to Money::multiply and Money::divide
- 2014-03-22 Allow RoundingMode to be specified when converting currencies
- 2014-03-22 CurrencyPair has an equals() method
- 2013-10-13 Base currency and counter currency in CurrencyPair named correctly.
- 2013-01-08 Removed the Doctrine2\MoneyType helper, to be replaced by something better in the future. It's available
             at https://gist.github.com/4485025 in case you need it.
- 2013-01-08 Use vendor/autoload.php instead of lib/bootstrap.php (or use PSR-0 autolaoding)
- 2012-12-10 Renamed Money::getUnits() to Money::getAmount()

tests/MoneyTest.php
<?php

/**
 * This file is part of the Money library.
 *
 * Copyright (c) 2011-2014 Mathias Verraes
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Money;

use InvalidArgumentException;
use UnexpectedValueException;

/**
 * @coversDefaultClass Money\Money
 * @uses Money\Currency
 * @uses Money\Money
 * @uses Money\CurrencyPair
 */
class MoneyTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @covers ::__callStatic
     */
    public function testFactoryMethods()
    {
        $this->assertEquals(
            Money::EUR(25),
            Money::EUR(10)->add(Money::EUR(15))
        );
        $this->assertEquals(
            Money::USD(25),
            Money::USD(10)->add(Money::USD(15))
        );
    }

    /**
     * @covers ::getAmount
     * @covers ::getCurrency
     */
    public function testGetters()
    {
        $m = new Money(100, $euro = new Currency('EUR'));
        $this->assertEquals(100, $m->getAmount());
        $this->assertEquals($euro, $m->getCurrency());
    }

    /**
     * @covers ::__construct
     * @expectedException InvalidArgumentException
     */
    public function testDecimalsThrowException()
    {
        $money = new Money(0.01, new Currency('EUR'));
    }

    /**
     * @covers ::__construct
     * @expectedException InvalidArgumentException
     */
    public function testStringThrowsException()
    {
        $money = new Money('100', new Currency('EUR'));
    }

    /**
     * @covers ::equals
     */
    public function testEquality()
    {
        $m1 = new Money(100, new Currency('EUR'));
        $m2 = new Money(100, new Currency('EUR'));
        $m3 = new Money(100, new Currency('USD'));
        $m4 = new Money(50, new Currency('EUR'));

        $this->assertTrue($m1->equals($m2));
        $this->assertFalse($m1->equals($m3));
        $this->assertFalse($m1->equals($m4));
    }

    public function testAddition()
    {
        $m1 = new Money(100, new Currency('EUR'));
        $m2 = new Money(100, new Currency('EUR'));
        $sum = $m1->add($m2);
        $expected = new Money(200, new Currency('EUR'));

        $this->assertEquals($expected, $sum);

        // Should return a new instance
        $this->assertNotSame($sum, $m1);
        $this->assertNotSame($sum, $m2);
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testDifferentCurrenciesCannotBeAdded()
    {
        $m1 = new Money(100, new Currency('EUR'));
        $m2 = new Money(100, new Currency('USD'));
        $m1->add($m2);
    }

    /**
     * @expectedException UnexpectedValueException
     */
    public function testResultNotAnInteger()
    {
        $m1 = new Money(PHP_INT_MAX, new Currency('EUR'));
        $m2 = new Money(1, new Currency('EUR'));
        $m1->add($m2);
    }

    /**
     * @covers ::subtract
     */
    public function testSubtraction()
    {
        $m1 = new Money(100, new Currency('EUR'));
        $m2 = new Money(200, new Currency('EUR'));
        $diff = $m1->subtract($m2);
        $expected = new Money(-100, new Currency('EUR'));

        $this->assertEquals($expected, $diff);

        // Should return a new instance
        $this->assertNotSame($diff, $m1);
        $this->assertNotSame($diff, $m2);
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testDifferentCurrenciesCannotBeSubtracted()
    {
        $m1 = new Money(100, new Currency('EUR'));
        $m2 = new Money(100, new Currency('USD'));
        $m1->subtract($m2);
    }

    public function testMultiplication()
    {
        $m = new Money(1, new Currency('EUR'));
        $this->assertEquals(
            new Money(2, new Currency('EUR')),
            $m->multiply(1.5)
        );
        $this->assertEquals(
            new Money(1, new Currency('EUR')),
            $m->multiply(1.5, Money::ROUND_HALF_DOWN)
        );

        $this->assertNotSame($m, $m->multiply(2));
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testInvalidMultiplicationOperand()
    {
        $m = new Money(1, new Currency('EUR'));
        $m->multiply('operand');
    }

    /**
     * @covers ::assertRoundingMode
     * @expectedException InvalidArgumentException
     */
    public function testInvalidRoundingMode()
    {
        $m = new Money(1, new Currency('EUR'));
        $m->multiply(1.2345, 'ROUNDING_MODE');
    }

    /**
     * @expectedException OverflowException
     */
    public function testMultiplicationOverflow()
    {
        $m = new Money(PHP_INT_MAX, new Currency('EUR'));
        $m->multiply(2);
    }

    /**
     * @expectedException UnderflowException
     */
    public function testMultiplicationUnderflow()
    {
        $m = new Money(~PHP_INT_MAX, new Currency('EUR'));
        $m->multiply(2);
    }

    /**
     * @covers ::divide
     */
    public function testDivision()
    {
        $m = new Money(10, new Currency('EUR'));
        $this->assertEquals(
            new Money(3, new Currency('EUR')),
            $m->divide(3)
        );
        $this->assertEquals(
            new Money(2, new Currency('EUR')),
            $m->divide(4, Money::ROUND_HALF_EVEN)
        );
        $this->assertEquals(
            new Money(3, new Currency('EUR')),
            $m->divide(3, Money::ROUND_HALF_ODD)
        );

        $this->assertNotSame($m, $m->divide(2));
    }

    public function testComparison()
    {
        $euro1 = new Money(1, new Currency('EUR'));
        $euro2 = new Money(2, new Currency('EUR'));
        $usd = new Money(1, new Currency('USD'));

        $this->assertTrue($euro2->greaterThan($euro1));
        $this->assertFalse($euro1->greaterThan($euro2));
        $this->assertTrue($euro1->lessThan($euro2));
        $this->assertFalse($euro2->lessThan($euro1));

        $this->assertEquals(-1, $euro1->compare($euro2));
        $this->assertEquals(1, $euro2->compare($euro1));
        $this->assertEquals(0, $euro1->compare($euro1));
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testDifferentCurrenciesCannotBeCompared()
    {
        Money::EUR(1)->compare(Money::USD(1));
    }

    /**
     * @covers ::allocate
     */
    public function testAllocation()
    {
        $m = new Money(100, new Currency('EUR'));
        list($part1, $part2, $part3) = $m->allocate(array(1, 1, 1));
        $this->assertEquals(new Money(34, new Currency('EUR')), $part1);
        $this->assertEquals(new Money(33, new Currency('EUR')), $part2);
        $this->assertEquals(new Money(33, new Currency('EUR')), $part3);

        $m = new Money(101, new Currency('EUR'));
        list($part1, $part2, $part3) = $m->allocate(array(1, 1, 1));
        $this->assertEquals(new Money(34, new Currency('EUR')), $part1);
        $this->assertEquals(new Money(34, new Currency('EUR')), $part2);
        $this->assertEquals(new Money(33, new Currency('EUR')), $part3);
    }

    /**
     * @covers ::allocate
     */
    public function testAllocationOrderIsImportant()
    {
        $m = new Money(5, new Currency('EUR'));
        list($part1, $part2) = $m->allocate(array(3, 7));
        $this->assertEquals(new Money(2, new Currency('EUR')), $part1);
        $this->assertEquals(new Money(3, new Currency('EUR')), $part2);

        $m = new Money(5, new Currency('EUR'));
        list($part1, $part2) = $m->allocate(array(7, 3));
        $this->assertEquals(new Money(4, new Currency('EUR')), $part1);
        $this->assertEquals(new Money(1, new Currency('EUR')), $part2);
    }

    /**
     * @covers ::allocateTo
     */
    public function testAllocationTo($value='')
    {
        $m = new Money(15, new Currency('EUR'));
        list($part1, $part2) = $m->allocateTo(2);
        $this->assertEquals(new Money(8, new Currency('EUR')), $part1);
        $this->assertEquals(new Money(7, new Currency('EUR')), $part2);

        $m = new Money(10, new Currency('EUR'));
        list($part1, $part2) = $m->allocateTo(2);
        $this->assertEquals(new Money(5, new Currency('EUR')), $part1);
        $this->assertEquals(new Money(5, new Currency('EUR')), $part2);
    }

    /**
     * @covers ::allocateTo
     * @expectedException InvalidArgumentException
     * @expectedExceptionMessage Number of targets must be an integer
     */
    public function testAllocationToInvalidTargets()
    {
        $m = new Money(15, new Currency('EUR'));
        $m->allocateTo('target');
    }

    /**
     * @covers ::isZero
     * @covers ::isNegative
     * @covers ::isPositive
     */
    public function testComparators()
    {
        $this->assertTrue(Money::EUR(0)->isZero());
        $this->assertTrue(Money::EUR(-1)->isNegative());
        $this->assertTrue(Money::EUR(1)->isPositive());
        $this->assertFalse(Money::EUR(1)->isZero());
        $this->assertFalse(Money::EUR(1)->isNegative());
        $this->assertFalse(Money::EUR(-1)->isPositive());
    }

    public static function provideStrings()
    {
        return array(
            array("1000", 100000),
            array("1000.0", 100000),
            array("1000.00", 100000),
            array("0.01", 1),
            array("1", 100),
            array("-1000", -100000),
            array("-1000.0", -100000),
            array("-1000.00", -100000),
            array("-0.01", -1),
            array("-1", -100),
            array("+1000", 100000),
            array("+1000.0", 100000),
            array("+1000.00", 100000),
            array("+0.01", 1),
            array("+1", 100),
            array(".99", 99),
            array("-.99", -99),
        );
    }

    /**
     * @covers ::stringToAmount
     * @dataProvider provideStrings
     */
    public function testStringToAmount($string, $units)
    {
        $this->assertEquals($units, Money::stringToAmount($string));
    }

    /**
     * @covers ::stringToAmount
     * @expectedException InvalidArgumentException
     */
    public function testCannotConvertStringToAmount()
    {
        Money::stringToAmount('THIS_IS_NOT_CONVERTABLE_TO_UNIT');
    }
}


src/Currency.php
<?php

/**
 * This file is part of the Money library.
 *
 * Copyright (c) 2011-2014 Mathias Verraes
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Money;

/**
 * Currency Value Object
 *
 * Holds Currency specific data
 *
 * @author Mathias Verraes
 */
class Currency
{
    /**
     * Currency code
     *
     * @var string
     */
    private $code;

    /**
     * @param string $code
     */
    public function __construct($code)
    {
        if (!is_string($code)) {
            throw new \InvalidArgumentException('Curency code should be string');
        }

        $this->code = $code;
    }

    /**
     * Returns the currency code
     *
     * @return string
     *
     * @deprecated Use getCode() instead
     */
    public function getName()
    {
        return $this->code;
    }

    /**
     * Returns the currency code
     *
     * @return string
     */
    public function getCode()
    {
        return $this->code;
    }

    /**
     * Checks whether this currency is the same as an other
     *
     * @param Currency $other
     *
     * @return boolean
     */
    public function equals(Currency $other)
    {
        return $this->code === $other->code;
    }

    /**
     * Checks whether this currency is available in the passed context
     *
     * @param AvailableCurrencies $currencies
     *
     * @return boolean
     */
    public function isAvailableWithin(AvailableCurrencies $currencies)
    {
        return $currencies->exists($this);
    }

    /**
     * @return string
     */
    public function __toString()
    {
        return $this->getCode();
    }
}


src/currencies.php
<?php
// From http://openexchangerates.org/currencies.json
// @codeCoverageIgnoreStart
return array(
	"AED" => "United Arab Emirates Dirham",
	"AFN" => "Afghan Afghani",
	"ALL" => "Albanian Lek",
	"AMD" => "Armenian Dram",
	"ANG" => "Netherlands Antillean Guilder",
	"AOA" => "Angolan Kwanza",
	"ARS" => "Argentine Peso",
	"AUD" => "Australian Dollar",
	"AWG" => "Aruban Florin",
	"AZN" => "Azerbaijani Manat",
	"BAM" => "Bosnia-Herzegovina Convertible Mark",
	"BBD" => "Barbadian Dollar",
	"BDT" => "Bangladeshi Taka",
	"BGN" => "Bulgarian Lev",
	"BHD" => "Bahraini Dinar",
	"BIF" => "Burundian Franc",
	"BMD" => "Bermudan Dollar",
	"BND" => "Brunei Dollar",
	"BOB" => "Bolivian Boliviano",
	"BRL" => "Brazilian Real",
	"BSD" => "Bahamian Dollar",
	"BTN" => "Bhutanese Ngultrum",
	"BWP" => "Botswanan Pula",
	"BYR" => "Belarusian Ruble",
	"BZD" => "Belize Dollar",
	"CAD" => "Canadian Dollar",
	"CDF" => "Congolese Franc",
	"CHF" => "Swiss Franc",
	"CLF" => "Chilean Unit of Account (UF)",
	"CLP" => "Chilean Peso",
	"CNY" => "Chinese Yuan",
	"COP" => "Colombian Peso",
	"CRC" => "Costa Rican Colón",
	"CUP" => "Cuban Peso",
	"CVE" => "Cape Verdean Escudo",
	"CZK" => "Czech Republic Koruna",
	"DJF" => "Djiboutian Franc",
	"DKK" => "Danish Krone",
	"DOP" => "Dominican Peso",
	"DZD" => "Algerian Dinar",
	"EEK" => "Estonian Kroon",
	"EGP" => "Egyptian Pound",
	"ETB" => "Ethiopian Birr",
	"EUR" => "Euro",
	"FJD" => "Fijian Dollar",
	"FKP" => "Falkland Islands Pound",
	"GBP" => "British Pound Sterling",
	"GEL" => "Georgian Lari",
	"GHS" => "Ghanaian Cedi",
	"GIP" => "Gibraltar Pound",
	"GMD" => "Gambian Dalasi",
	"GNF" => "Guinean Franc",
	"GTQ" => "Guatemalan Quetzal",
	"GYD" => "Guyanaese Dollar",
	"HKD" => "Hong Kong Dollar",
	"HNL" => "Honduran Lempira",
	"HRK" => "Croatian Kuna",
	"HTG" => "Haitian Gourde",
	"HUF" => "Hungarian Forint",
	"IDR" => "Indonesian Rupiah",
	"ILS" => "Israeli New Sheqel",
	"INR" => "Indian Rupee",
	"IQD" => "Iraqi Dinar",
	"IRR" => "Iranian Rial",
	"ISK" => "Icelandic Króna",
	"JEP" => "Jersey Pound",
	"JMD" => "Jamaican Dollar",
	"JOD" => "Jordanian Dinar",
	"JPY" => "Japanese Yen",
	"KES" => "Kenyan Shilling",
	"KGS" => "Kyrgystani Som",
	"KHR" => "Cambodian Riel",
	"KMF" => "Comorian Franc",
	"KPW" => "North Korean Won",
	"KRW" => "South Korean Won",
	"KWD" => "Kuwaiti Dinar",
	"KYD" => "Cayman Islands Dollar",
	"KZT" => "Kazakhstani Tenge",
	"LAK" => "Laotian Kip",
	"LBP" => "Lebanese Pound",
	"LKR" => "Sri Lankan Rupee",
	"LRD" => "Liberian Dollar",
	"LSL" => "Lesotho Loti",
	"LTL" => "Lithuanian Litas",
	"LVL" => "Latvian Lats",
	"LYD" => "Libyan Dinar",
	"MAD" => "Moroccan Dirham",
	"MDL" => "Moldovan Leu",
	"MGA" => "Malagasy Ariary",
	"MKD" => "Macedonian Denar",
	"MMK" => "Myanma Kyat",
	"MNT" => "Mongolian Tugrik",
	"MOP" => "Macanese Pataca",
	"MRO" => "Mauritanian Ouguiya",
	"MUR" => "Mauritian Rupee",
	"MVR" => "Maldivian Rufiyaa",
	"MWK" => "Malawian Kwacha",
	"MXN" => "Mexican Peso",
	"MYR" => "Malaysian Ringgit",
	"MZN" => "Mozambican Metical",
	"NAD" => "Namibian Dollar",
	"NGN" => "Nigerian Naira",
	"NIO" => "Nicaraguan Córdoba",
	"NOK" => "Norwegian Krone",
	"NPR" => "Nepalese Rupee",
	"NZD" => "New Zealand Dollar",
	"OMR" => "Omani Rial",
	"PAB" => "Panamanian Balboa",
	"PEN" => "Peruvian Nuevo Sol",
	"PGK" => "Papua New Guinean Kina",
	"PHP" => "Philippine Peso",
	"PKR" => "Pakistani Rupee",
	"PLN" => "Polish Zloty",
	"PYG" => "Paraguayan Guarani",
	"QAR" => "Qatari Rial",
	"RON" => "Romanian Leu",
	"RSD" => "Serbian Dinar",
	"RUB" => "Russian Ruble",
	"RUR" => "Russian Ruble",
	"RWF" => "Rwandan Franc",
	"SAR" => "Saudi Riyal",
	"SBD" => "Solomon Islands Dollar",
	"SCR" => "Seychellois Rupee",
	"SDG" => "Sudanese Pound",
	"SEK" => "Swedish Krona",
	"SGD" => "Singapore Dollar",
	"SHP" => "Saint Helena Pound",
	"SLL" => "Sierra Leonean Leone",
	"SOS" => "Somali Shilling",
	"SRD" => "Surinamese Dollar",
	"STD" => "São Tomé and Príncipe Dobra",
	"SVC" => "Salvadoran Colón",
	"SYP" => "Syrian Pound",
	"SZL" => "Swazi Lilangeni",
	"THB" => "Thai Baht",
	"TJS" => "Tajikistani Somoni",
	"TMT" => "Turkmenistani Manat",
	"TND" => "Tunisian Dinar",
	"TOP" => "Tongan Paʻanga",
	"TRY" => "Turkish Lira",
	"TTD" => "Trinidad and Tobago Dollar",
	"TWD" => "New Taiwan Dollar",
	"TZS" => "Tanzanian Shilling",
	"UAH" => "Ukrainian Hryvnia",
	"UGX" => "Ugandan Shilling",
	"USD" => "United States Dollar",
	"UYU" => "Uruguayan Peso",
	"UZS" => "Uzbekistan Som",
	"VEF" => "Venezuelan Bolívar",
	"VND" => "Vietnamese Dong",
	"VUV" => "Vanuatu Vatu",
	"WST" => "Samoan Tala",
	"XAF" => "CFA Franc BEAC",
	"XCD" => "East Caribbean Dollar",
	"XDR" => "Special Drawing Rights",
	"XOF" => "CFA Franc BCEAO",
	"XPF" => "CFP Franc",
	"YER" => "Yemeni Rial",
	"ZAR" => "South African Rand",
	"ZMK" => "Zambian Kwacha",
	"ZWL" => "Zimbabwean Dollar",
);
// @codeCoverageIgnoreEnd


src/Money.php
<?php

/**
 * This file is part of the Money library.
 *
 * Copyright (c) 2011-2014 Mathias Verraes
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Money;

use InvalidArgumentException;
use OverflowException;
use UnderflowException;
use UnexpectedValueException;

/**
 * Money Value Object
 *
 * @author Mathias Verraes
 */
class Money
{
    const ROUND_HALF_UP   = PHP_ROUND_HALF_UP;
    const ROUND_HALF_DOWN = PHP_ROUND_HALF_DOWN;
    const ROUND_HALF_EVEN = PHP_ROUND_HALF_EVEN;
    const ROUND_HALF_ODD  = PHP_ROUND_HALF_ODD;


    /**
     * Internal value
     *
     * @var int
     */
    private $amount;


    /**
     * @var Currency
     */
    private $currency;


    /**
     * @param int  $amount   Amount, expressed in the smallest units of $currency (eg cents)
     * @param Currency $currency
     *
     * @throws InvalidArgumentException If amount is not integer
     */
    public function __construct($amount, Currency $currency)
    {
        if(!is_int($amount)) {
            throw new InvalidArgumentException('Amount must be an integer');
        }

        $this->amount = $amount;
        $this->currency = $currency;
    }


    /**
     * Convenience factory method for a Money object
     *
     * <code>
     * $fiveDollar = Money::USD(500);
     * </code>
     *
     * @param string $method
     * @param array  $arguments
     *
     * @return Money
     */
    public static function __callStatic($method, $arguments)
    {
        return new Money($arguments[0], new Currency($method));
    }





    /**
     * Checkers area
     *
     */


    /**
     * Checks whether a Money has the same Currency as this
     *
     * @param Money $other
     *
     * @return boolean
     */
    public function isSameCurrency(Money $other)
    {
        return $this->currency->equals($other->currency);
    }


    /**
     * Checks if the value represented by this object is zero
     *
     * @return boolean
     */
    public function isZero()
    {
        return $this->amount === 0;
    }


    /**
     * Checks if the value represented by this object is positive
     *
     * @return boolean
     */
    public function isPositive()
    {
        return $this->amount > 0;
    }


    /**
     * Checks if the value represented by this object is negative
     *
     * @return boolean
     */
    public function isNegative()
    {
        return $this->amount < 0;
    }


    /**
     * Checks whether the value represented by this object equals to the other
     *
     * @param Money $other
     *
     * @return boolean
     */
    public function equals(Money $other)
    {
        return $this->isSameCurrency($other) && $this->amount == $other->amount;
    }


    /**
     * Checks whether the value represented by this object is greater than the other
     *
     * @param Money $other
     *
     * @return boolean
     */
    public function greaterThan(Money $other)
    {
        return 1 == $this->compare($other);
    }


    /**
     * @param \Money\Money $other
     * @return bool
     */
    public function greaterThanOrEqual(Money $other)
    {
        return 0 >= $this->compare($other);
    }


    /**
     * Checks whether the value represented by this object is less than the other
     *
     * @param Money $other
     *
     * @return boolean
     */
    public function lessThan(Money $other)
    {
        return -1 == $this->compare($other);
    }


    /**
     * @param \Money\Money $other
     * @return bool
     */
    public function lessThanOrEqual(Money $other)
    {
        return 0 <= $this->compare($other);
    }


    /**
     * Returns an integer less than, equal to, or greater than zero
     * if the value of this object is considered to be respectively
     * less than, equal to, or greater than the other
     *
     * @param Money $other
     *
     * @return int
     */
    public function compare(Money $other)
    {
        $this->assertSameCurrency($other);

        if($this->amount < $other->amount) {
            return -1;
        } elseif($this->amount == $other->amount) {
            return 0;
        } else {
            return 1;
        }
    }





    /**
     * Getters area
     *
     */


    /**
     * Returns the value represented by this object
     *
     * @return int
     */
    public function getAmount()
    {
        return $this->amount;
    }


    /**
     * Возвращает значение getAmount() / 100
     *
     * @param  mixed $roundingMode [false | Money::ROUND_HALF_DOWN | Money::ROUND_HALF_EVEN | Money::ROUND_HALF_ODD | Money::ROUND_HALF_UP]
     *                             Смотри в ман по PHP, функция round, для информации о ключах
     *                             Если false, то без округления
     * @return int|float
     */
    public function getAmountNormal($roundingMode = self::ROUND_HALF_UP)
    {
        return ($roundingMode ? round($this->amount / 100, 0, $roundingMode) : $this->amount / 100);
    }


    /**
     * Returns the currency of this object
     *
     * @return Currency
     */
    public function getCurrency()
    {
        return $this->currency;
    }


    /**
     * Возвращает "человеческое" представление: 34 105,23 руб
     * Для удобного визуального тестирования
     *
     * @param  integer $decimals Кол-во знаков после запятой
     *
     * @return String
     */
    public function getHuman($decimals = 2)
    {
        $currency = (string)$this->getCurrency();

        return number_format($this->getAmount() / 100, $decimals, ',', ' ') . (!empty($currency) ? ' '. $currency : '' );
    }





    /**
     * Operations area
     *
     */


    /**
     * Returns a new Money object that represents
     * the sum of this and an other Money object
     *
     * @param Money $addend
     *
     * @return Money
     */
    public function add(Money $addend)
    {
        $this->assertSameCurrency($addend);

        $amount = $this->amount + $addend->amount;

        $this->assertInteger($amount);

        return $this->newInstance($amount);
    }


    /**
     * Returns a new Money object that represents
     * the difference of this and an other Money object
     *
     * @param Money $subtrahend
     *
     * @return Money
     */
    public function subtract(Money $subtrahend)
    {
        $this->assertSameCurrency($subtrahend);

        $amount = $this->amount - $subtrahend->amount;

        $this->assertInteger($amount);

        return $this->newInstance($amount);
    }


    /**
     * Returns a new Money object that represents
     * the multiplied value by the given factor
     *
     * @param numeric $multiplier
     * @param int $roundingMode
     *
     * @return Money
     */
    public function multiply($multiplier, $roundingMode = self::ROUND_HALF_UP)
    {
        $this->assertOperand($multiplier);

        $this->assertRoundingMode($roundingMode);

        $product = round($this->amount * $multiplier, 0, $roundingMode);

        $product = $this->castInteger($product);

        return $this->newInstance($product);
    }


    /**
     * Returns a new Money object that represents
     * the divided value by the given factor
     *
     * @param numeric $divisor
     * @param int $roundingMode
     *
     * @return Money
     */
    public function divide($divisor, $roundingMode = self::ROUND_HALF_UP)
    {
        $this->assertOperand($divisor);

        $this->assertRoundingMode($roundingMode);

        $quotient = round($this->amount / $divisor, 0, $roundingMode);

        $quotient = $this->castInteger($quotient);

        return $this->newInstance($quotient);
    }


    /**
     * Конвертирует текущую валюту в $targetCurrency по курсу $conversionRate
     * @param Currency $targetCurrency
     * @param float|int $conversionRate
     * @param int $roundingMode
     * @return Money
     */
    public function convert(Currency $targetCurrency, $conversionRate, $roundingMode = Money::ROUND_HALF_UP)
    {
        $this->assertRoundingMode($roundingMode);
        $amount = round($this->amount * $conversionRate, 0, $roundingMode);
        $amount = $this->castInteger($amount);
        return new Money($amount, $targetCurrency);
    }


    /**
     * Распределить деньги по долям
     * Usage:
     *     list($my_cut, $investors_cut) = Money::EUR(5)->allocate(70, 30);
     *
     * @param array $ratios
     *
     * @return Money[]
     */
    public function allocate(array $ratios)
    {
        $remainder = $this->amount;
        $results = array();
        $total = array_sum($ratios);

        foreach ($ratios as $ratio) {
            $share = $this->castInteger($this->amount * $ratio / $total);
            $results[] = $this->newInstance($share);
            $remainder -= $share;
        }

        for ($i = 0; $remainder > 0; $i++) {
            $results[$i]->amount++;
            $remainder--;
        }

        return $results;
    }


    /**
     * Распределить деньги поровну на N частей
     *
     * @param int $n
     *
     * @return Money[]
     *
     * @throws InvalidArgumentException If number of targets is not an integer
     */
    public function allocateTo($n)
    {
        if(!is_int($n)) {
            throw new InvalidArgumentException('Number of targets must be an integer');
        }

        $amount = intval($this->amount / $n);
        $results = array();

        for ($i = 0; $i < $n; $i++) {
            $results[$i] = $this->newInstance($amount);
        }

        for ($i = 0; $i < $this->amount % $n; $i++) {
            $results[$i]->amount++;
        }

        return $results;
    }


    /**
     * Creates amounts from string
     *
     * @param string $string Something like "2939.00"
     *
     * @return int
     *
     * @throws InvalidArgumentException If $string cannot be parsed
     */
    public static function stringToAmount($string)
    {
        $sign = "(?P<sign>[-\+])?";
        $digits = "(?P<digits>\d*)";
        $separator = "(?P<separator>[.,])?";
        $decimals = "(?P<decimal1>\d)?(?P<decimal2>\d)?";
        $pattern = "/^".$sign.$digits.$separator.$decimals."$/";

        if(!preg_match($pattern, trim($string), $matches)) {
            throw new InvalidArgumentException("The value could not be parsed as money");
        }

        $units = $matches['sign'] == "-" ? "-" : "";
        $units .= $matches['digits'];
        $units .= isset($matches['decimal1']) ? $matches['decimal1'] : "0";
        $units .= isset($matches['decimal2']) ? $matches['decimal2'] : "0";

        return (int) $units;
    }






    /**
     * Private area
     *
     */


    /**
     * Returns a new Money instance based on the current one using the Currency
     *
     * @param int $amount
     *
     * @return Money
     */
    private function newInstance($amount)
    {
        return new Money($amount, $this->currency);
    }


    /**
     * Asserts that a Money has the same currency as this
     *
     * @throws InvalidArgumentException If $other has a different currency
     */
    private function assertSameCurrency(Money $other)
    {
        if(!$this->isSameCurrency($other)) {
            throw new InvalidArgumentException('Currencies must be identical');
        }
    }


    /**
     * Asserts that integer remains integer after arithmetic operations
     *
     * @param  numeric $amount
     */
    private function assertInteger($amount)
    {
        if(!is_int($amount)) {
            throw new UnexpectedValueException('The result of arithmetic operation is not an integer');
        }
    }


    /**
     * Asserts that the operand is integer or float
     *
     * @throws InvalidArgumentException If $operand is neither integer nor float
     */
    private function assertOperand($operand)
    {
        if(!is_int($operand) && !is_float($operand)) {
            throw new InvalidArgumentException('Operand should be an integer or a float');
        }
    }


    /**
     * Asserts that an integer value didn't become something else
     * (after some arithmetic operation)
     *
     * @param numeric $amount
     *
     * @throws OverflowException If integer overflow occured
     * @throws UnderflowException If integer underflow occured
     */
    private function assertIntegerBounds($amount)
    {
        if($amount > PHP_INT_MAX) {
            throw new OverflowException;
        } elseif($amount < ~PHP_INT_MAX) {
            throw new UnderflowException;
        }
    }


    /**
     * Casts an amount to integer ensuring that an overflow/underflow did not occur
     *
     * @param numeric $amount
     *
     * @return int
     */
    private function castInteger($amount)
    {
        $this->assertIntegerBounds($amount);

        return intval($amount);
    }


    /**
     * Asserts that rounding mode is a valid integer value
     *
     * @param int $roundingMode
     *
     * @throws InvalidArgumentException If $roundingMode is not valid
     */
    private function assertRoundingMode($roundingMode)
    {
        if(!in_array(
            $roundingMode,
            array(self::ROUND_HALF_DOWN, self::ROUND_HALF_EVEN, self::ROUND_HALF_ODD, self::ROUND_HALF_UP)
        )) {
            throw new InvalidArgumentException(
                'Rounding mode should be Money::ROUND_HALF_DOWN | ' .
                'Money::ROUND_HALF_EVEN | Money::ROUND_HALF_ODD | ' .
                'Money::ROUND_HALF_UP'
            );
        }
    }

}
